openapi/workflow_execution_service.swagger.yaml

```diff
@@ -100,15 +100,36 @@ paths:
         - WorkflowExecutionService
     post:
       summary: |-
-        Run a workflow, this endpoint will allow you to create a new workflow request and
-        retrieve its tracking ID to monitor its progress.  An important assumption in this
-        endpoint is that the workflow_params JSON will include parameterizations along with
-        input and output files.  The latter two may be on S3, Google object storage, local filesystems,
-        etc.  This specification makes no distinction.  However, it is assumed that the submitter
-        is using URLs that this system both understands and can access. For Amazon S3, this could
-        be accomplished by given the credentials associated with a WES service access to a
-        particular bucket.  The details are important for a production system and user on-boarding
-        but outside the scope of this spec.
+        Run a workflow.  This endpoint creates a new workflow run and
+        returns the workflow ID to monitor its progress.
+
+        The workflow_params JSON object specifies input parameters,
+        such as input files.  The exact format of the JSON object
+        depends on the conventions of the workflow language being
+        used.  Input files should be addressed by URL.  The WES
+        endpoint must understand and be able to access URLs supplied
+        in the input.  This is implementation specific.
+
+        This endpoint can be used two different ways, selected by
```

Comment:
This is not how I understood our agreement.

I was picturing the submission request to be a multipart with:

- The JSON bundle of most things
- 0 or more workflow descriptor files

@mckinsel Since you & I have the action item to design this, I'm curious to get your read on this. Was I the outlier here?

Reply:
I apologize for jumping in on your action item.  I was doing some prototyping so I wanted to write up my findings, and thought a PR would be the most productive way to kickstart the discussion.

We could specify that it only accepts multipart, but in any case one of the "parts" needs to be the json request, and the other "parts" are the files needed to run the workflow.  The part containing the json payload can either be the first item, or an item with a particular name.  I went with a particular name ("body") but it is fine either way.

Selecting on content-type makes life simpler for the submitter when they want to provide a workflow_url and not upload files.  But that case can also be accommodated with multipart with a single item ("body").

Reply:
Hah no I was happy you did :) It was also good because it made clear much faster that there was a difference of opinion on what was decided.

The second paragraph is exactly how I had been picturing it. 

The third paragraph works for me insofar as the JSON blob has a well specified form field when multipart

My preference would be for the second paragraph, but I tend to dislike overloading things in general so I recognize that this quite possibly just my bias showing.

Reply:
Just to be completely clear, I mean choosing behavior based on Content-Type of the primary headers as to whether the body of the request is just the request json or whether it is a multipart stream, and not talking about Content-Type fields on the individual parts.

But in the interests of keeping the interface minimal, I can take out that part, we specify that:

* it always has to be multipart
* the request json is the part named "body"
* even if the submitter doesn't need to attach files, they still have to use multipart encoding on the body part

Reply:
We should, at a minimum, wait for Marcus to comment prior to changing the PR

Reply:
Hey apologies for the delay in response. I like the multipart-only idea just in terms of keeping the implementation simple for now. Does it make sense to just include the other workflow request fields as form values? I'm not sure what we get out of json encoding and then multipart encoding.

Also, I don't think there's a way to specify this interface in swagger 2? @geoffjentry I think this is the issue that's led to the WorkflowParams{1..N} in the cromwell API, right?

Reply:
> Hey apologies for the delay in response. I like the multipart-only idea just in terms of keeping the implementation simple for now.

I'm okay with making this simplification, but also foresee the next person looking at the spec wondering why they need to do a multipart encode when all they want to do is execute a workflow by URL.

> Does it make sense to just include the other workflow request fields as form values?   I'm not sure what we get out of json encoding and then multipart encoding.

The json payload is just one of the parts (the rest are the files).  Technically you could pass each field as part as well, but having the payload be json encoded seems more consistent with the rest of the API.

Reply:
@tetron "but also foresee the next person looking at the spec wondering why" FWIW when I have shown the spec to folks locally they're always wondering why the JSON blob approach is being used and it's not a multipart form altogether. I say this not as pushing for what @mckinsel suggested but rather to note that people are going to be coming to this with a variety of biases and will by nature be expecting to see what they're used to seeing.

@mckinsel Yes, there is (was?) a limitation in Swagger where you can't have arbitrarily large set of fields with the same name. That WorkflowParams{1..N} is purely a swagger thing, it's not necessary in e.g. curl, but was the only way we could get it to work at all in swagger.

Reply:
I guess some of the non-`workflow_descriptor` fields are going to be JSON blobs either way: `workflow_params` and `workflow_engine_parameters`. So clients and implementers are doing some nested encoding/decoding in either scenario.

The swagger limitation appears to be fixed in openapi 3. You can do something like this (I think):
```yaml
content:
  multipart/form-data:
    schema:
      WorkflowRequest:
        type: object
        properties:
          workflow_descriptor:
            type: array
            items:
              type: string
              format: binary
```

Reply:
If there's a preferred encoding in OpenAPI 3 then we should probably adopt that.

Reply:
@mckinsel On your first point I don't think that's right. IIRC those were both specified to be actual JSON objects (instead of the current "whatever the heck one wants, but we know it'll be JSON" which means they can be sub-objects in the main JSON blob. The same would be true of form fields (as we see in the Cromwell API). So I don't think either side should require escaping?